<a href="https://colab.research.google.com/github/anmol-iisc/psa_dspi_rmdn12/blob/main/Assignment2_1_26Sep_0926PM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Assignment 2: Exploratory Data Analysis on Diabetes Dataset

## Background
Diabetes is one of the most prevalent chronic diseases and creates significant burdens on healthcare systems. The **Diabetes 130-US hospitals dataset** contains over **100,000 records** of hospital admissions for diabetic patients across **130 U.S. hospitals (1999–2008)**. It includes more than **50 attributes**, covering demographics, admission details, diagnoses, lab results, medications, and hospital outcomes.  

Like many real-world healthcare datasets, it presents challenges such as missing values, categorical codes (ICD-9 diagnoses, drug prescriptions), imbalances in outcome variables (e.g., readmission), and inconsistent data formats. Conducting an in-depth **Exploratory Data Analysis (EDA)** will help uncover insights into patient demographics, treatment outcomes, and factors influencing readmission rates.  

---

## Objectives
The goal of this assignment is to conduct a comprehensive **qualitative and quantitative EDA** of the Diabetes dataset. You will:  

- **Understand** the dataset structure, including variables, data types, and overall composition.  
- **Detect and address** data quality issues, such as missing values, duplicates, and inconsistent categories.  
- **Conduct univariate, bivariate, and multivariate analysis** using both qualitative and quantitative EDA, supported by visualization techniques (bar charts, count plots, stacked charts, heatmaps) and statistical methods (descriptive statistics, distribution analysis, and exploration of variable relationships).  
- **Visualize** patterns and correlations to uncover key insights and trends.  

---

## Submission Instructions
- This is an **individual assignment** and carries **5% of the total marks** of this course.
- You should not discuss your progress or solutions with anyone else.  
- Use this **template Jupyter Notebook and dataset** for your work.  

At the beginning of the notebook, clearly state:  
1. **Your Name and Roll Number**  
2. **Python environment details** (version, libraries used).  
3. Whether you are running locally or on **Google Colab** (*preferred, to ensure all students work in a consistent environment*).  

Additional guidelines:  
- Ensure that your notebook is **fully reproducible**: every cell must run sequentially without errors to regenerate the results.  
- Submit a **single completed notebook file (.ipynb)** before the deadline.  

---


## Dataset Source
- Source of Data : https://archive.ics.uci.edu/dataset/296/diabetes+130-us+hospitals+for+years+1999-2008
- It represents **hospital admissions of diabetic patients** across 130 U.S. hospitals from 1999 to 2008.
- Read the **Variables Table** in the link above, which provides details of individual variables such as type, role, and description.

##  Dataset Overview - "Diabetes_dataset.csv"
- **Number of records:** ~101,766 hospital admissions  
- **Number of features:** 53  
- **Unit of analysis:** Each row represents **one hospital admission** for a patient (a patient can have multiple admissions).  
- **Target variable (Goal):** Predict whether a patient is **readmitted within 30 days** of discharge.  

###  Key Features with Mappings
- **Admission Type**: `admission_type_id` → mapped to `admission_type_desc`  
- **Discharge Disposition**: `discharge_disposition_id` → mapped to `discharge_desc`  
- **Admission Source**: `admission_source_id` → mapped to `admission_source_desc`  

---

##  Important Notes
- **Missing values:** Several features contain missing values, often recorded as `"?"` (not as `NaN`).   
- **ID columns:** Both numeric IDs and their human-readable descriptions are included after merging. IDs may be useful for re-checking or remapping, while descriptions are helpful for analysis.  


**Note:** Do not just run plots — write down **interpretations** of your findings.  
The goal is to practice *thinking about the data*, not just generating charts.


# 📌 Preamble: Environment Details

Before starting the assignment, please fill in the following:

- **Name**: Anmol Gupta  
- **Email**: anmolg@iisc.ac.in  
- **Environment**: Google Colab  

Run the code cell below to print your Python and library versions.


In [931]:
# Preamble: Environment Check

import sys
import pandas as pd
import numpy as np
import matplotlib
import seaborn as sns

print("Python version:", sys.version)
print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)
print("Matplotlib version:", matplotlib.__version__)
print("Seaborn version:", sns.__version__)

Python version: 3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0]
Pandas version: 2.2.2
NumPy version: 2.0.2
Matplotlib version: 3.10.0
Seaborn version: 0.13.2


## Load and Inspect the dataset structure

#### Load the dataset

In [932]:
df=pd.read_csv("Diabetes_dataset.csv")

#### Inspect the dataset

In [933]:
df.shape

(101766, 53)

In [934]:
df.head()

encounter_id  patient_nbr             race  gender      age weight  \
0       2278392      8222157        Caucasian  Female   [0-10)      ?   
1        149190     55629189        Caucasian  Female  [10-20)      ?   
2         64410     86047875  AfricanAmerican  Female  [20-30)      ?   
3        500364     82442376        Caucasian    Male  [30-40)      ?   
4         16680     42519267        Caucasian    Male  [40-50)      ?   

   admission_type_id  discharge_disposition_id  admission_source_id  \
0                  6                        25                    1   
1                  1                         1                    7   
2                  1                         1                    7   
3                  1                         1                    7   
4                  1                         1                    7   

   time_in_hospital  ... glipizide-metformin glimepiride-pioglitazone  \
0                 1  ...                  No                       No   
1                 3  ...                  No                       No   
2                 2  ...                  No                       No   
3                 2  ...                  No                       No   
4                 1  ...                  No                       No   

   metformin-rosiglitazone  metformin-pioglitazone  change  diabetesMed  \
0                       No                      No      No           No   
1                       No                      No      Ch          Yes   
2                       No                      No      No          Yes   
3                       No                      No      Ch          Yes   
4                       No                      No      Ch          Yes   

   readmitted  admission_type_desc      discharge_desc admission_source_desc  
0          NO                  NaN          Not Mapped    Physician Referral  
1         >30            Emergency  Discharged to home        Emergency Room  
2          NO            Emergency  Discharged to home        Emergency Room  
3          NO            Emergency  Discharged to home        Emergency Room  
4          NO            Emergency  Discharged to home        Emergency Room  

[5 rows x 53 columns]

In [935]:
df.describe()

encounter_id   patient_nbr  admission_type_id  \
count  1.017660e+05  1.017660e+05      101766.000000   
mean   1.652016e+08  5.433040e+07           2.024006   
std    1.026403e+08  3.869636e+07           1.445403   
min    1.252200e+04  1.350000e+02           1.000000   
25%    8.496119e+07  2.341322e+07           1.000000   
50%    1.523890e+08  4.550514e+07           1.000000   
75%    2.302709e+08  8.754595e+07           3.000000   
max    4.438672e+08  1.895026e+08           8.000000   

       discharge_disposition_id  admission_source_id  time_in_hospital  \
count             101766.000000        101766.000000     101766.000000   
mean                   3.715642             5.754437          4.395987   
std                    5.280166             4.064081          2.985108   
min                    1.000000             1.000000          1.000000   
25%                    1.000000             1.000000          2.000000   
50%                    1.000000             7.000000          4.000000   
75%                    4.000000             7.000000          6.000000   
max                   28.000000            25.000000         14.000000   

       num_lab_procedures  num_procedures  num_medications  number_outpatient  \
count       101766.000000   101766.000000    101766.000000      101766.000000   
mean            43.095641        1.339730        16.021844           0.369357   
std             19.674362        1.705807         8.127566           1.267265   
min              1.000000        0.000000         1.000000           0.000000   
25%             31.000000        0.000000        10.000000           0.000000   
50%             44.000000        1.000000        15.000000           0.000000   
75%             57.000000        2.000000        20.000000           0.000000   
max            132.000000        6.000000        81.000000          42.000000   

       number_emergency  number_inpatient  number_diagnoses  
count     101766.000000     101766.000000     101766.000000  
mean           0.197836          0.635566          7.422607  
std            0.930472          1.262863          1.933600  
min            0.000000          0.000000          1.000000  
25%            0.000000          0.000000          6.000000  
50%            0.000000          0.000000          8.000000  
75%            0.000000          1.000000          9.000000  
max           76.000000         21.000000         16.000000

In [936]:
df.describe().T

count          mean           std      min  \
encounter_id              101766.0  1.652016e+08  1.026403e+08  12522.0   
patient_nbr               101766.0  5.433040e+07  3.869636e+07    135.0   
admission_type_id         101766.0  2.024006e+00  1.445403e+00      1.0   
discharge_disposition_id  101766.0  3.715642e+00  5.280166e+00      1.0   
admission_source_id       101766.0  5.754437e+00  4.064081e+00      1.0   
time_in_hospital          101766.0  4.395987e+00  2.985108e+00      1.0   
num_lab_procedures        101766.0  4.309564e+01  1.967436e+01      1.0   
num_procedures            101766.0  1.339730e+00  1.705807e+00      0.0   
num_medications           101766.0  1.602184e+01  8.127566e+00      1.0   
number_outpatient         101766.0  3.693572e-01  1.267265e+00      0.0   
number_emergency          101766.0  1.978362e-01  9.304723e-01      0.0   
number_inpatient          101766.0  6.355659e-01  1.262863e+00      0.0   
number_diagnoses          101766.0  7.422607e+00  1.933600e+00      1.0   

                                 25%          50%           75%          max  
encounter_id              84961194.0  152388987.0  2.302709e+08  443867222.0  
patient_nbr               23413221.0   45505143.0  8.754595e+07  189502619.0  
admission_type_id                1.0          1.0  3.000000e+00          8.0  
discharge_disposition_id         1.0          1.0  4.000000e+00         28.0  
admission_source_id              1.0          7.0  7.000000e+00         25.0  
time_in_hospital                 2.0          4.0  6.000000e+00         14.0  
num_lab_procedures              31.0         44.0  5.700000e+01        132.0  
num_procedures                   0.0          1.0  2.000000e+00          6.0  
num_medications                 10.0         15.0  2.000000e+01         81.0  
number_outpatient                0.0          0.0  0.000000e+00         42.0  
number_emergency                 0.0          0.0  0.000000e+00         76.0  
number_inpatient                 0.0          0.0  1.000000e+00         21.0  
number_diagnoses                 6.0          8.0  9.000000e+00         16.0

In [937]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 53 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   encounter_id              101766 non-null  int64 
 1   patient_nbr               101766 non-null  int64 
 2   race                      101766 non-null  object
 3   gender                    101766 non-null  object
 4   age                       101766 non-null  object
 5   weight                    101766 non-null  object
 6   admission_type_id         101766 non-null  int64 
 7   discharge_disposition_id  101766 non-null  int64 
 8   admission_source_id       101766 non-null  int64 
 9   time_in_hospital          101766 non-null  int64 
 10  payer_code                101766 non-null  object
 11  medical_specialty         101766 non-null  object
 12  num_lab_procedures        101766 non-null  int64 
 13  num_procedures            101766 non-null  int64 
 14  num_

## Data Pre-processing (cleaning, transformation, handling missing values, etc.)

We can see that there are 53 columns in the original dataframe.

To make our job easier while performing analysis, let us break it down into separate (dummy) dataframes based on the column types.

##### Numerical Columns

In [938]:
numeric_df = df.select_dtypes(include='number')

In [939]:
numeric_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 13 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   encounter_id              101766 non-null  int64
 1   patient_nbr               101766 non-null  int64
 2   admission_type_id         101766 non-null  int64
 3   discharge_disposition_id  101766 non-null  int64
 4   admission_source_id       101766 non-null  int64
 5   time_in_hospital          101766 non-null  int64
 6   num_lab_procedures        101766 non-null  int64
 7   num_procedures            101766 non-null  int64
 8   num_medications           101766 non-null  int64
 9   number_outpatient         101766 non-null  int64
 10  number_emergency          101766 non-null  int64
 11  number_inpatient          101766 non-null  int64
 12  number_diagnoses          101766 non-null  int64
dtypes: int64(13)
memory usage: 10.1 MB


In [940]:
for col in numeric_df.columns:
    print(col, "has", numeric_df[col].duplicated().sum(), "duplicate values")

encounter_id has 0 duplicate values
patient_nbr has 30248 duplicate values
admission_type_id has 101758 duplicate values
discharge_disposition_id has 101740 duplicate values
admission_source_id has 101749 duplicate values
time_in_hospital has 101752 duplicate values
num_lab_procedures has 101648 duplicate values
num_procedures has 101759 duplicate values
num_medications has 101691 duplicate values
number_outpatient has 101727 duplicate values
number_emergency has 101733 duplicate values
number_inpatient has 101745 duplicate values
number_diagnoses has 101750 duplicate values


Since `encounter_id` has all unique values, it would not contribute towards identifying any relationship.

Hence, we can drop it from the original dataframe.

**NOTE**: We use the `numeric_df` only for analysis. Any operation we perform (like dropping a column) is performed on the ***orginal dataframe*** itself. Only for maintaining consistency and to avoid confusion later, we perform the same operation on the the dummy `numeric_df` also. The same would apply to other dummy dataframes that we create later too.

In [941]:
df.drop(columns=['encounter_id'], inplace=True)
# Doing the same on the dummy dataframe for consistency purpose
numeric_df.drop(columns=['encounter_id'], inplace=True)

In [942]:
numeric_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 12 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   patient_nbr               101766 non-null  int64
 1   admission_type_id         101766 non-null  int64
 2   discharge_disposition_id  101766 non-null  int64
 3   admission_source_id       101766 non-null  int64
 4   time_in_hospital          101766 non-null  int64
 5   num_lab_procedures        101766 non-null  int64
 6   num_procedures            101766 non-null  int64
 7   num_medications           101766 non-null  int64
 8   number_outpatient         101766 non-null  int64
 9   number_emergency          101766 non-null  int64
 10  number_inpatient          101766 non-null  int64
 11  number_diagnoses          101766 non-null  int64
dtypes: int64(12)
memory usage: 9.3 MB


Let us check if any of our numerical columns has '?'

In [943]:
(numeric_df == '?').sum()

patient_nbr                 0
admission_type_id           0
discharge_disposition_id    0
admission_source_id         0
time_in_hospital            0
num_lab_procedures          0
num_procedures              0
num_medications             0
number_outpatient           0
number_emergency            0
number_inpatient            0
number_diagnoses            0
dtype: int64

Let us do a high-level analysis of the remaining columns in numeric_df.

In [944]:
for col in numeric_df.columns:
    print('================================================')
    print('\033[1m'+col+'\033[0m')
    print('Number of unique values:',numeric_df[col].nunique(),"\n")
    print('Unique Values:',numeric_df[col].unique(),"\n\n")


patient_nbr
Number of unique values: 71518 

Unique Values: [  8222157  55629189  86047875 ... 140199494 120975314 175429310] 


admission_type_id
Number of unique values: 8 

Unique Values: [6 1 2 3 4 5 8 7] 


discharge_disposition_id
Number of unique values: 26 

Unique Values: [25  1  3  6  2  5 11  7 10  4 14 18  8 13 12 16 17 22 23  9 20 15 24 28
 19 27] 


admission_source_id
Number of unique values: 17 

Unique Values: [ 1  7  2  4  5  6 20  3 17  8  9 14 10 22 11 25 13] 


time_in_hospital
Number of unique values: 14 

Unique Values: [ 1  3  2  4  5 13 12  9  7 10  6 11  8 14] 


num_lab_procedures
Number of unique values: 118 

Unique Values: [ 41  59  11  44  51  31  70  73  68  33  47  62  60  55  49  75  45  29
  35  42  66  36  19  64  25  53  52  87  27  37  46  28  48  72  10   2
  65  67  40  54  58  57  43  32  83  34  39  69  38  56  22  96  78  61
  88  50   1  18  82   9  63  24  71  77  81  76  90  93   3 103  13  80
  85  16  15  12  30  23  17  21  79  26   5  9

Some of these columns can benefit from binning - but let us look at it once we come to univariate analysis.
Other than that, from a data cleaning point-of-view, we are done with the numercial columns.

We will come back to the numerical columns later, if needed.

##### Object Type Columns

In [945]:
object_df = df.select_dtypes(include='object')

In [946]:
object_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 40 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   race                      101766 non-null  object
 1   gender                    101766 non-null  object
 2   age                       101766 non-null  object
 3   weight                    101766 non-null  object
 4   payer_code                101766 non-null  object
 5   medical_specialty         101766 non-null  object
 6   diag_1                    101766 non-null  object
 7   diag_2                    101766 non-null  object
 8   diag_3                    101766 non-null  object
 9   max_glu_serum             5346 non-null    object
 10  A1Cresult                 17018 non-null   object
 11  metformin                 101766 non-null  object
 12  repaglinide               101766 non-null  object
 13  nateglinide               101766 non-null  object
 14  chlo

We still have 40 columns as object data type.

So let us try to split them further.

In [947]:
object_df.columns

Index(['race', 'gender', 'age', 'weight', 'payer_code', 'medical_specialty',
       'diag_1', 'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult', 'metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'admission_type_desc', 'discharge_desc', 'admission_source_desc'],
      dtype='object')

In [948]:
drug_columns = ['metformin',
       'repaglinide', 'nateglinide', 'chlorpropamide', 'glimepiride',
       'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone']

In [949]:
non_drug_columns = ['race', 'gender', 'age', 'weight', 'payer_code', 'medical_specialty',
       'diag_1', 'diag_2', 'diag_3', 'max_glu_serum', 'A1Cresult','change', 'diabetesMed', 'readmitted',
       'admission_type_desc', 'discharge_desc', 'admission_source_desc']

In [950]:
drug_df = object_df[drug_columns]
drug_df.columns.nunique()

23

In [951]:
non_drug_df = object_df[non_drug_columns]
non_drug_df.columns.nunique()

17

###### Let us start with the drug columns.

In [952]:
df.shape

(101766, 52)

In [953]:
for col in drug_df.columns:
    print(col, "has", drug_df[col].nunique(), "unique values")

metformin has 4 unique values
repaglinide has 4 unique values
nateglinide has 4 unique values
chlorpropamide has 4 unique values
glimepiride has 4 unique values
acetohexamide has 2 unique values
glipizide has 4 unique values
glyburide has 4 unique values
tolbutamide has 2 unique values
pioglitazone has 4 unique values
rosiglitazone has 4 unique values
acarbose has 4 unique values
miglitol has 4 unique values
troglitazone has 2 unique values
tolazamide has 3 unique values
examide has 1 unique values
citoglipton has 1 unique values
insulin has 4 unique values
glyburide-metformin has 4 unique values
glipizide-metformin has 2 unique values
glimepiride-pioglitazone has 2 unique values
metformin-rosiglitazone has 2 unique values
metformin-pioglitazone has 2 unique values


We can see that `examide` and `citoglipton` have only one unique value. So we can drop them from the original dataframe.

In [954]:
df.drop(columns=['examide', 'citoglipton'], inplace=True)
# Doing the same on the dummy dataframe for consistency purpose
drug_df.drop(columns=['examide', 'citoglipton'], inplace=True)

/tmp/ipython-input-3699290549.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  drug_df.drop(columns=['examide', 'citoglipton'], inplace=True)


In [955]:
drug_df.isnull().sum()

metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide                  0
insulin                     0
glyburide-metformin         0
glipizide-metformin         0
glimepiride-pioglitazone    0
metformin-rosiglitazone     0
metformin-pioglitazone      0
dtype: int64

In [956]:
(drug_df == '?').sum()

metformin                   0
repaglinide                 0
nateglinide                 0
chlorpropamide              0
glimepiride                 0
acetohexamide               0
glipizide                   0
glyburide                   0
tolbutamide                 0
pioglitazone                0
rosiglitazone               0
acarbose                    0
miglitol                    0
troglitazone                0
tolazamide                  0
insulin                     0
glyburide-metformin         0
glipizide-metformin         0
glimepiride-pioglitazone    0
metformin-rosiglitazone     0
metformin-pioglitazone      0
dtype: int64

In [957]:
for col in drug_df.columns:
    print('================================================')
    print('\033[1m'+col+'\033[0m')
    print('Number of unique values:',drug_df[col].nunique(),"\n")
    print('Unique Values:',drug_df[col].unique(),"\n")
    print('Number of null values:',drug_df[col].isnull().sum(),"\n\n")

metformin
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Up' 'Down'] 

Number of null values: 0 


repaglinide
Number of unique values: 4 

Unique Values: ['No' 'Up' 'Steady' 'Down'] 

Number of null values: 0 


nateglinide
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Down' 'Up'] 

Number of null values: 0 


chlorpropamide
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Down' 'Up'] 

Number of null values: 0 


glimepiride
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Down' 'Up'] 

Number of null values: 0 


acetohexamide
Number of unique values: 2 

Unique Values: ['No' 'Steady'] 

Number of null values: 0 


glipizide
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Up' 'Down'] 

Number of null values: 0 


glyburide
Number of unique values: 4 

Unique Values: ['No' 'Steady' 'Up' 'Down'] 

Number of null values: 0 


tolbutamide
Number of unique values: 2 

Unique Values: ['No' 'Steady'] 

Number of null values: 0 



From data cleaning, point-of-view, these columns look fine. None of them have any null values, and all of them have between two to four categorical values.

###### Let us continue with the non-drug columns.

In [958]:
for col in non_drug_df.columns:
    print(col, "has", non_drug_df[col].duplicated().sum(), "duplicate values")

race has 101760 duplicate values
gender has 101763 duplicate values
age has 101756 duplicate values
weight has 101756 duplicate values
payer_code has 101748 duplicate values
medical_specialty has 101693 duplicate values
diag_1 has 101049 duplicate values
diag_2 has 101017 duplicate values
diag_3 has 100976 duplicate values
max_glu_serum has 101762 duplicate values
A1Cresult has 101762 duplicate values
change has 101764 duplicate values
diabetesMed has 101764 duplicate values
readmitted has 101763 duplicate values
admission_type_desc has 101758 duplicate values
discharge_desc has 101740 duplicate values
admission_source_desc has 101749 duplicate values


In [959]:
(non_drug_df == '?').sum()

race                      2273
gender                       0
age                          0
weight                   98569
payer_code               40256
medical_specialty        49949
diag_1                      21
diag_2                     358
diag_3                    1423
max_glu_serum                0
A1Cresult                    0
change                       0
diabetesMed                  0
readmitted                   0
admission_type_desc          0
discharge_desc               0
admission_source_desc        0
dtype: int64

Replacing all '?' with null values.

In [960]:
df.replace("?", np.nan, inplace=True)
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df.replace("?", np.nan, inplace=True)

/tmp/ipython-input-876314273.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_drug_df.replace("?", np.nan, inplace=True)


In [961]:
non_drug_df.isnull().sum()

race                      2273
gender                       0
age                          0
weight                   98569
payer_code               40256
medical_specialty        49949
diag_1                      21
diag_2                     358
diag_3                    1423
max_glu_serum            96420
A1Cresult                84748
change                       0
diabetesMed                  0
readmitted                   0
admission_type_desc       5291
discharge_desc            3691
admission_source_desc     6781
dtype: int64

In [962]:
percent_missing = non_drug_df.isnull().sum()*100/len(df)
percent_missing_sorted = percent_missing.sort_values(ascending=False)
percent_missing_sorted

weight                   96.858479
max_glu_serum            94.746772
A1Cresult                83.277322
medical_specialty        49.082208
payer_code               39.557416
admission_source_desc     6.663326
admission_type_desc       5.199182
discharge_desc            3.626948
race                      2.233555
diag_3                    1.398306
diag_2                    0.351787
diag_1                    0.020636
gender                    0.000000
age                       0.000000
change                    0.000000
diabetesMed               0.000000
readmitted                0.000000
dtype: float64

Columns `weight`, `max_glu_serum`, `A1Cresult` have very high null value percentages, so we can drop them from the original dataframe.

In [963]:
df.drop(columns=['weight', 'max_glu_serum', 'A1Cresult'], inplace=True)
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df.drop(columns=['weight', 'max_glu_serum', 'A1Cresult'], inplace=True)

/tmp/ipython-input-2845509147.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_drug_df.drop(columns=['weight', 'max_glu_serum', 'A1Cresult'], inplace=True)


Columns `medical_specialty` and `payer_code` also have a significant percentage of null values, but let us analyze them a bit more.

`medical_specialty`

In [964]:
non_drug_df['medical_specialty'].unique()

array(['Pediatrics-Endocrinology', nan, 'InternalMedicine',
       'Family/GeneralPractice', 'Cardiology', 'Surgery-General',
       'Orthopedics', 'Gastroenterology',
       'Surgery-Cardiovascular/Thoracic', 'Nephrology',
       'Orthopedics-Reconstructive', 'Psychiatry', 'Emergency/Trauma',
       'Pulmonology', 'Surgery-Neuro',
       'Obsterics&Gynecology-GynecologicOnco', 'ObstetricsandGynecology',
       'Pediatrics', 'Hematology/Oncology', 'Otolaryngology',
       'Surgery-Colon&Rectal', 'Pediatrics-CriticalCare', 'Endocrinology',
       'Urology', 'Psychiatry-Child/Adolescent', 'Pediatrics-Pulmonology',
       'Neurology', 'Anesthesiology-Pediatric', 'Radiology',
       'Pediatrics-Hematology-Oncology', 'Psychology', 'Podiatry',
       'Gynecology', 'Oncology', 'Pediatrics-Neurology',
       'Surgery-Plastic', 'Surgery-Thoracic',
       'Surgery-PlasticwithinHeadandNeck', 'Ophthalmology',
       'Surgery-Pediatric', 'Pediatrics-EmergencyMedicine',
       'PhysicalMedicineandRe

Let us keep `medical_specialty` for now, and impute the missing values as a new value 'Unknown'.

In [965]:
df['medical_specialty'] = df['medical_specialty'].fillna('Unknown')
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df['medical_specialty'] = non_drug_df['medical_specialty'].fillna('Unknown')

/tmp/ipython-input-2342112168.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_drug_df['medical_specialty'] = non_drug_df['medical_specialty'].fillna('Unknown')


`payer_code`

In [966]:
df['payer_code'].unique()

array([nan, 'MC', 'MD', 'HM', 'UN', 'BC', 'SP', 'CP', 'SI', 'DM', 'CM',
       'CH', 'PO', 'WC', 'OT', 'OG', 'MP', 'FR'], dtype=object)

Let us remove `payer_code`, as anyway it should not contribute to readmission, along wiht having a large percentage of null values.  

In [967]:
df.drop(columns=['payer_code'], inplace=True)
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df.drop(columns=['payer_code'], inplace=True)

/tmp/ipython-input-1142169840.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_drug_df.drop(columns=['payer_code'], inplace=True)


Let us do a high-level analysis of the remaining columns in non_drug_df.

In [968]:
for col in non_drug_df.columns:
    print('================================================')
    print('\033[1m'+col+'\033[0m')
    print('Number of unique values:',non_drug_df[col].nunique(),"\n")
    print('Unique Values:',non_drug_df[col].unique(),"\n")
    print('Number of null values:',non_drug_df[col].isnull().sum(),"\n\n")

race
Number of unique values: 5 

Unique Values: ['Caucasian' 'AfricanAmerican' nan 'Other' 'Asian' 'Hispanic'] 

Number of null values: 2273 


gender
Number of unique values: 3 

Unique Values: ['Female' 'Male' 'Unknown/Invalid'] 

Number of null values: 0 


age
Number of unique values: 10 

Unique Values: ['[0-10)' '[10-20)' '[20-30)' '[30-40)' '[40-50)' '[50-60)' '[60-70)'
 '[70-80)' '[80-90)' '[90-100)'] 

Number of null values: 0 


medical_specialty
Number of unique values: 73 

Unique Values: ['Pediatrics-Endocrinology' 'Unknown' 'InternalMedicine'
 'Family/GeneralPractice' 'Cardiology' 'Surgery-General' 'Orthopedics'
 'Gastroenterology' 'Surgery-Cardiovascular/Thoracic' 'Nephrology'
 'Orthopedics-Reconstructive' 'Psychiatry' 'Emergency/Trauma'
 'Pulmonology' 'Surgery-Neuro' 'Obsterics&Gynecology-GynecologicOnco'
 'ObstetricsandGynecology' 'Pediatrics' 'Hematology/Oncology'
 'Otolaryngology' 'Surgery-Colon&Rectal' 'Pediatrics-CriticalCare'
 'Endocrinology' 'Urology' 'Psychiatr

Let us pick these up one by one.

`race`

Imputing all null values with the value 'Unknown'

In [969]:
print(df['race'].isnull().sum())

2273


In [970]:
df['race'] = df['race'].fillna('Unknown')
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df['race'] = non_drug_df['race'].fillna('Unknown')

/tmp/ipython-input-2112190087.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_drug_df['race'] = non_drug_df['race'].fillna('Unknown')


In [971]:
print(df['race'].isnull().sum())

0


In [972]:
df['race'].unique()

array(['Caucasian', 'AfricanAmerican', 'Unknown', 'Other', 'Asian',
       'Hispanic'], dtype=object)

`gender`

In [973]:
non_drug_df['gender'].value_counts()

gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64

In 2025, it is iniquitous to disregard non-binary genders. However, since the dataset contains only 3 rows with 'Unknown/Invalid' - as compared to 101,763 rows with Female/Male - we would not be able to derive any meaningful analysis from keeping these rows. Hence, we drop these rows.

In [974]:
df = df.drop(df[df['gender'] == 'Unknown/Invalid'].index)
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df = non_drug_df.drop(non_drug_df[non_drug_df['gender'] == 'Unknown/Invalid'].index)

In [975]:
df['gender'].value_counts()

gender
Female    54708
Male      47055
Name: count, dtype: int64

In [976]:
df.shape

(101763, 46)

`age`

In [977]:
non_drug_df['age'].value_counts()

age
[70-80)     26066
[60-70)     22482
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64

Seems fine, so keeping it as is.

`medical_specialty`

In [978]:
non_drug_df['medical_specialty'].value_counts()

medical_specialty
Unknown                   49947
InternalMedicine          14635
Emergency/Trauma           7565
Family/GeneralPractice     7440
Cardiology                 5351
                          ...  
Proctology                    1
Speech                        1
SportsMedicine                1
Perinatology                  1
Neurophysiology               1
Name: count, Length: 73, dtype: int64

Seems fine for now, so keeping it as is.

`diag_1`, `diag_2`, `diag_3`

These three columns are the primary, secondary and additional secondary diagnosis respectively, coded as first three digits of ICD9.

Let us transform them into their equivalent categories.

Source: Wikipedia

The following is a list of codes for International Statistical Classification of Diseases and Related Health Problems.



*   List of ICD-9 codes 001-139: infectious and parasitic diseases
*   List of ICD-9 codes 140-239: neoplasms
*   List of ICD-9 codes 240-279: endocrine, nutritional and metabolic diseases, and immunity disorders
*   List of ICD-9 codes 280-289: diseases of the blood and blood-forming organs
*   List of ICD-9 codes 290-319: mental disorders
*   List of ICD-9 codes 320-389: diseases of the nervous system and sense organs
*   List of ICD-9 codes 390-459: diseases of the circulatory system
*   List of ICD-9 codes 460-519: diseases of the respiratory system
*   List of ICD-9 codes 520-579: diseases of the digestive system
*   List of ICD-9 codes 580-629: diseases of the genitourinary system
*   List of ICD-9 codes 630-679: complications of pregnancy, childbirth, and the puerperium
*   List of ICD-9 codes 680-709: diseases of the skin and subcutaneous tissue
*   List of ICD-9 codes 710-739: diseases of the musculoskeletal system and connective tissue
*   List of ICD-9 codes 740-759: congenital anomalies
*   List of ICD-9 codes 760-779: certain conditions originating in the perinatal period
*   List of ICD-9 codes 780-799: symptoms, signs, and ill-defined conditions
*   List of ICD-9 codes 800-999: injury and poisoning
*   List of ICD-9 codes E and V codes: external causes of injury and supplemental classification

Firstly, let us impute the missing values with '-1'.

In [979]:
for col in ['diag_1', 'diag_2', 'diag_3']:
  print(col,non_drug_df[col].isnull().sum())

diag_1 21
diag_2 358
diag_3 1423


In [980]:
for col in ['diag_1', 'diag_2', 'diag_3']:
    df[col] = df[col].fillna('-1')
    # Doing the same on the dummy dataframe for consistency purpose
    non_drug_df[col] = non_drug_df[col].fillna('-1')

In [981]:
for col in ['diag_1', 'diag_2', 'diag_3']:
  print(col,df[col].isnull().sum())

diag_1 0
diag_2 0
diag_3 0


Now, let us transform the numbers into categories, based on the ICD-9 Code.

In [982]:
def transform_diag(icd9_code):
    if icd9_code.lower().startswith(('e', 'v')):
        category = 'External Causes of Injury'
        return category
    else:
        value = float(icd9_code)

        if value>=1 and value<=139:
            category = 'Infectious and Parasitic Diseases'
        elif value>=140 and value<=239:
            category = 'Neoplasms'
        elif value>=240 and value<=279:
            category = 'Endocrine, Nutritional and Metabolic Diseases, and Immunity Disorders'
        elif value>=280 and value<=289:
            category = 'Diseases of the Blood and Blood-Forming Organs'
        elif value>=290 and value<=319:
            category = 'Mental Disorders'
        elif value>=320 and value<=389:
            category = 'Diseases of Nervous System and Sense Organs'
        elif value>=390 and value<=459:
            category = 'Diseases of Circulatory System'
        elif value>=460 and value<=519:
            category = 'Diseases of Respiratory System'
        elif value>=520 and value<=579:
            category = 'Diseases of Digestive System'
        elif value>=580 and value<=629:
            category = 'Diseases of Genitourinary System'
        elif value>=630 and value<=679:
            category = 'Complications of Pregnancy, Childbirth, and Puerperium'
        elif value>=680 and value<=709:
            category = 'Diseases of Skin and Subcutaneous Tissue'
        elif value>=710 and value<=739:
            category = 'Diseases of Musculoskeletal System and Connective Tissue'
        elif value>=740 and value<=759:
            category = 'Congenital Anomalies'
        elif value>=760 and value<=779:
            category = 'Diseases of Perinatal Period'
        elif value>=780 and value<=799:
            category = 'Symptoms, Signs, and Ill-defined Conditions'
        elif value>=800 and value<=999:
            category = 'Injury and Poisoning'
        elif value == -1:
            category = 'Unknown'
        else: #Should not enter this as we have covered all scenarios
            category = 'Other'

        return category

In [983]:
for col in ['diag_1', 'diag_2', 'diag_3']:
  df[col] = df[col].apply(transform_diag)
  # Doing the same on the dummy dataframe for consistency purpose
  non_drug_df[col] = non_drug_df[col].apply(transform_diag)

In [984]:
for col in ['diag_1', 'diag_2', 'diag_3']:
    print('================================================')
    print('\033[1m'+col+'\033[0m')
    print('Number of unique values:',df[col].nunique(),"\n")
    print('Unique Values:',df[col].unique(),"\n")

diag_1
Number of unique values: 18 

Unique Values: ['Endocrine, Nutritional and Metabolic Diseases, and Immunity Disorders'
 'Complications of Pregnancy, Childbirth, and Puerperium'
 'Infectious and Parasitic Diseases' 'Neoplasms'
 'Diseases of Circulatory System' 'Diseases of Respiratory System'
 'Injury and Poisoning' 'Diseases of Skin and Subcutaneous Tissue'
 'Diseases of Musculoskeletal System and Connective Tissue'
 'Diseases of Digestive System' 'External Causes of Injury'
 'Symptoms, Signs, and Ill-defined Conditions'
 'Diseases of Genitourinary System' 'Mental Disorders'
 'Diseases of Nervous System and Sense Organs'
 'Diseases of the Blood and Blood-Forming Organs' 'Unknown'
 'Congenital Anomalies'] 

diag_2
Number of unique values: 18 

Unique Values: ['Unknown'
 'Endocrine, Nutritional and Metabolic Diseases, and Immunity Disorders'
 'Neoplasms' 'Diseases of Circulatory System'
 'Diseases of Respiratory System'
 'Diseases of the Blood and Blood-Forming Organs' 'Injury and 

`change`

This is already a categorical column with no null values, and two unique values ('No' and 'Ch'). So keeping as is.

`diabetesMed`

This is already a categorical column with no null values, and two unique values ('No' and 'Yes'). So keeping as is.

`admission_type_desc`, `discharge_desc`, `admission_source_desc`

These are already categorical columns.

First, let us impute the null values.

Imputing all null values with the value 'Unknown'

In [985]:
for col in ['admission_type_desc', 'discharge_desc', 'admission_source_desc']:
  print(col,df[col].isnull().sum())

admission_type_desc 5291
discharge_desc 3691
admission_source_desc 6781


In [986]:
for col in ['admission_type_desc', 'discharge_desc', 'admission_source_desc']:
  df[col] = df[col].fillna('Unknown')
  # Doing the same on the dummy dataframe for consistency purpose
  non_drug_df[col] = non_drug_df[col].fillna('Unknown')

In [987]:
for col in ['admission_type_desc', 'discharge_desc', 'admission_source_desc']:
  print(col,df[col].isnull().sum())

admission_type_desc 0
discharge_desc 0
admission_source_desc 0


Let us look at the columns of `discharge_desc` again.

In [988]:
non_drug_df['discharge_desc'].unique()

array(['Not Mapped', 'Discharged to home',
       'Discharged/transferred to SNF',
       'Discharged/transferred to home with home health service',
       'Discharged/transferred to another short term hospital',
       'Discharged/transferred to another type of inpatient care institution',
       'Expired', 'Left AMA',
       'Neonate discharged to another hospital for neonatal aftercare',
       'Discharged/transferred to ICF', 'Hospice / medical facility',
       'Unknown',
       'Discharged/transferred to home under care of Home IV provider',
       'Hospice / home',
       'Still patient or expected to return for outpatient services',
       'Discharged/transferred/referred another institution for outpatient services',
       'Discharged/transferred/referred to this institution for outpatient services',
       'Discharged/transferred to another rehab fac including rehab units of a hospital .',
       'Discharged/transferred to a long term care hospital.',
       'Admitted as an i

We have a few columns where the patients have expired.

Unfortunately, they have no scope of getting re-admitted, so we drop those records.  

In [989]:
df = df.drop(df[df['discharge_desc'].str.contains('Expired', case=False)].index)
# Doing the same on the dummy dataframe for consistency purpose
non_drug_df = non_drug_df.drop(non_drug_df[non_drug_df['discharge_desc'].str.contains('Expired', case=False)].index)

In [990]:
df.shape

(100111, 46)

NOTE: Practically, even the patients who are advised Hospice care do not have a realistic chance at readmission.

However, the humane side of me believes in mircales - so I am not dropping those patient records.

`readmitted`

Finally, we come to our target variable - `readmitted`.

This column is already categorial, with three unique values.

However, to make things interesting, let us create another column `readmitted_binary` with just two values - 'Yes' and 'No'. We map '>30' and '<30' both to 'Yes'.

In [991]:
df['readmitted_binary'] = df['readmitted']
df['readmitted_binary'] = df['readmitted_binary'].str.replace('>30', 'Yes')
df['readmitted_binary'] = df['readmitted_binary'].str.replace('<30', 'Yes')
df['readmitted_binary'] = df['readmitted_binary'].str.replace('NO', 'No')

In [992]:
df['readmitted_binary'].unique()

array(['No', 'Yes'], dtype=object)

In [993]:
df.shape

(100111, 47)

In [994]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100111 entries, 0 to 101765
Data columns (total 47 columns):
 #   Column                    Non-Null Count   Dtype 
---  ------                    --------------   ----- 
 0   patient_nbr               100111 non-null  int64 
 1   race                      100111 non-null  object
 2   gender                    100111 non-null  object
 3   age                       100111 non-null  object
 4   admission_type_id         100111 non-null  int64 
 5   discharge_disposition_id  100111 non-null  int64 
 6   admission_source_id       100111 non-null  int64 
 7   time_in_hospital          100111 non-null  int64 
 8   medical_specialty         100111 non-null  object
 9   num_lab_procedures        100111 non-null  int64 
 10  num_procedures            100111 non-null  int64 
 11  num_medications           100111 non-null  int64 
 12  number_outpatient         100111 non-null  int64 
 13  number_emergency          100111 non-null  int64 
 14  number_in

In [899]:
for col in df.columns:
    #print(col, df[col].duplicated().sum(), df[col].duplicated().sum()/(len(df)-1)*100)
    # print number of unique values per column and sort them in ascending order
    print(col, df[col].nunique())

patient_nbr 70436
race 6
gender 2
age 10
admission_type_id 8
discharge_disposition_id 23
admission_source_id 17
time_in_hospital 14
medical_specialty 73
num_lab_procedures 118
num_procedures 7
num_medications 75
number_outpatient 39
number_emergency 33
number_inpatient 21
diag_1 18
diag_2 18
diag_3 18
number_diagnoses 16
metformin 4
repaglinide 4
nateglinide 4
chlorpropamide 4
glimepiride 4
acetohexamide 2
glipizide 4
glyburide 4
tolbutamide 2
pioglitazone 4
rosiglitazone 4
acarbose 4
miglitol 4
troglitazone 2
tolazamide 3
insulin 4
glyburide-metformin 4
glipizide-metformin 2
glimepiride-pioglitazone 2
metformin-rosiglitazone 2
metformin-pioglitazone 2
change 2
diabetesMed 2
readmitted 3
admission_type_desc 8
discharge_desc 23
admission_source_desc 17
readmitted_binary 2


In [900]:
# Dropping columns having same value across all rows
df.drop(columns=['examide', 'citoglipton'], inplace=True)

KeyError: "['examide', 'citoglipton'] not found in axis"

In [ ]:
df.shape

In [ ]:
# distinct values in readmitted column
df['readmitted'].value_counts()

In [ ]:
df['readmitted_label'] = df['readmitted'].map({'NO':0, '>30':1, '<30':2})

In [ ]:
df['readmitted_label'].unique()

In [ ]:
df.head()

In [ ]:
percent_missing = df.isnull().sum()*100/len(df)
percent_missing_sorted = percent_missing.sort_values(ascending=False)
percent_missing_sorted

In [ ]:
# plot on readmitted label -
sns.countplot(data=df, x='readmitted_label') # frequency - count

In [ ]:
df.isnull().sum()

In [ ]:
df.duplicated().sum()

In [ ]:
# boxplot of age with readmitted
sns.boxplot(x=df['readmitted'],y=df['age'])

In [ ]:
# countplot of Age group
sns.countplot(data=df, x='age')

In [ ]:
pd.crosstab(df['age'], df['readmitted'], margins=True)  # cross tabulation of Age_group and survived columns

In [ ]:
pd.crosstab(df['age'], df['readmitted_label'], margins=True)  # cross tabulation of Age_group and survived columns

In [ ]:
# boxplot of gender with readmitted
sns.boxplot(x=df['readmitted'],y=df['gender'])

In [ ]:
# countplot for gender with readmitted
sns.countplot(data=df, x = 'gender', hue='readmitted')

In [ ]:
pd.crosstab(df['gender'], df['readmitted_label'], margins=True)  # cross tabulation of Age_group and survived columns

In [ ]:
# compute readmitted by gender - group by gender
readmitted_by_gender = df.groupby('gender')['readmitted_label'].mean()
readmitted_by_gender

##NOTE: Does not make sense!

In [ ]:
sns.boxplot(df, x='race', y='age', palette='rainbow')

In [ ]:
df['admission_type_desc'].unique()

In [ ]:
df['discharge_desc'].unique()

In [ ]:
df['admission_source_desc'].unique()

## Univariate Analysis

## Bivariate Analysis

## Multivariate Analysis

## Summarize insights